packages/integration-sdk-runtime/src/logger/registerEventEmitters.ts

```diff
@@ -0,0 +1,91 @@
+import { IntegrationLogger } from '.';
+
+type EventType = 'multipleResolves'|'unhandledRejection'|'uncaughtException';
+
+type LifecycleErrorCallback = (err: Error, event: EventType) => void;
+
+function multipleResolveListener(callback: LifecycleErrorCallback) {
+  // conforms to type MultipleResolveListener, see node_modules/@types/node/globals.d.ts
+  return (type: 'resolve' | 'reject', promise: Promise<any>, value: any) => {
+    if (type === 'reject') {
+      callback(value, 'multipleResolves');
+    }
+  };
+}
+
+function unhandledRejectionListener(callback: LifecycleErrorCallback) {
+  // conforms to type UnhandledRejectionListener, see node_modules/@types/node/globals.d.ts
+  return (reason: {} | null | undefined, promise: Promise<any>) => {
+    callback(reason as Error, 'unhandledRejection');
+  };
+}
+
+function uncaughtExceptionListener(callback: LifecycleErrorCallback) {
+  // conforms to type UncaughtExceptionListener, see node_modules/@types/node/globals.d.ts
+  return (error: Error) => {
+    callback(error, 'uncaughtException');
+  };
+}
+
+/**
+ * Because individual integrations contain code in their `validateInvocation` and `integrationSteps`
+ * that could include unhandled rejections or multiple resolves, this function is exported as a convenience
+ * to catch and handle events emitted from an integration. This emitter should be instrumented as early
+ * as possible in the node process in order to handle any instrumentation exceptions.
+ *
+ * Before exiting the node process, unregister these event emitters using `unregisterEventEmitters`
+ */
+export function registerEventEmitters(callback: LifecycleErrorCallback) {
+  process.on('multipleResolves', multipleResolveListener(callback));
+  process.on('unhandledRejection', unhandledRejectionListener(callback));
+  process.on('uncaughtException', uncaughtExceptionListener(callback));
+}
+
+/**
+ * Call this function before exiting the node process when using `registerEventEmitters`
+ */
+export function unregisterEventEmitters(callback: LifecycleErrorCallback) {
+  process.nextTick(() => {
+    process.removeListener(
+      'multipleResolves',
+      multipleResolveListener(callback),
```

Comment:
I'm not sure that this will behave as you expect. `removeListener` requires a direct reference to the original function that was registered. This will create a new function instead.

Reply:
Great catch, Austin.

What do you think about this as an implementation?

```typescript
// first, define event emitter. maintains singleton & throws if defineEventEmitterCallback is called more than once
/**
 * First define event emitter. Maintains singleton & throws if defineEventEmitter is called
 * more than once
 */
defineEventEmitterCallback((err, event) => { logger.error({ err, event });

/**
 * registers based on the callback defined in defineEventEmitterCallback. throws if 
 * defineEventEmitterCallback has not been called
 */
registerEventEmitters();

// EXECUTION CODE HERE

/**
 * unregisters the event emitters based on the callback defined above. throws if 
 * registerEventEmitters was not called
 */
unregisterEventEmitters();
 ```

Alternatively, a slightly simpler implementation but one that requires `registerEventEmitters` to also maintain the callback singleton

```typescript
/**
 * Maintains a callback singleton & throws if registerEventEmitters was already called.
 */
registerEventEmitters((err, event) => { logger.error({ err, event });

// EXECUTION CODE HERE

/**
 * SAME AS ABOVE
 */
unregisterEventEmitters();

Reply:
__Discussed offline__

 I think `registerEventEmitters` should return the functions that were registered and the `unregisterEventEmitters` function should take the return value of `registerEventEmitters`.

Perhaps something like this:

```typescript
function registerEventEmitters(callback) {
  const multipleResolveListenerFn = multipleResolveListener(callback);
  // Register all the goodies
  return {
    'multipleResolves': multipleResolveListenerFn
  }
}
const registeredEventEmitters = registerEventEmitters(cb)
unregisterEventEmitters(registeredEventEmitters)
```